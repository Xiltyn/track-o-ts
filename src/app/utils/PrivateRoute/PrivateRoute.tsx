import * as React from 'react';
import { Redirect, Route, RouteComponentProps } from "react-router";
import { connect } from "react-redux";
import { RootState } from "../../modules";
import { AppLoader } from "../../components/__universal/AppLoader/AppLoader";

export namespace PrivateRoute {
    export interface Props extends RouteComponentProps<void> {
        isAuthenticated:boolean,
        isBusy:boolean,
        Component:React.ReactNode,
    }
}

@connect(
    (state:RootState):Pick<PrivateRoute.Props, 'isAuthenticated'|'isBusy'> => {
        return {
            isAuthenticated: (state.auth.user && state.auth.user.email) !== null,
            isBusy: state.auth.started !== undefined && state.auth.started,
        }
    }
)

export class PrivateRoute extends React.Component<any|PrivateRoute.Props> {
    static defaultProps:Pick<PrivateRoute.Props, 'isAuthenticated'|'isBusy'> = {
        isAuthenticated: false,
        isBusy: false,
    };

    render() {
        const {
            isAuthenticated,
            isBusy,
            Component,
            ...rest
        } = this.props;

        return (
            <Route { ...rest } render={ (props:any) => (
                isBusy
                    ? <AppLoader style={ {
                        width: 24,
                        height: 24,
                        margin: '-12px',
                    } } />
                    : isAuthenticated
                    ? <Component { ...props } />
                    : <Redirect to='/login'/>
            ) }/>
        )
    }
}
