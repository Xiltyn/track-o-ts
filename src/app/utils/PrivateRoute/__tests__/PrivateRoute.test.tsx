import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import configureStore, { MockStore } from 'redux-mock-store';
import { PrivateRoute } from "../PrivateRoute";

describe('<|:===== <PrivateRoute /> =====:|>', () => {
    const mockStore = configureStore();
    const initialState = {
        auth: {
            isAuthenticated: false,
            isBusy: false,
            user: {},
        }
    };
    let wrapper:ShallowWrapper<PrivateRoute.Props>;
    let store:MockStore;

    beforeEach(() => {
        // Initializes shallow render of tested Component with mocked redux store
        store = mockStore(initialState);
        wrapper = shallow(<PrivateRoute store={ store }/>);
    });

    it('renders component', () => {
        // Compares Component snapshot with previously taken snapshot to check for unexpected data changes
       expect(wrapper).toMatchSnapshot();
   });

    it('isAuthenticated matches initialState', () => {
        // Checks whether initialState.isAuthenticated matches Component.props
      expect(wrapper.prop('isAuthenticated')).toEqual(initialState.auth.isAuthenticated);
    });

    it('isBusy matches initialState', () => {
        // Checks whether initialState.isBusy matches Component.props
        expect(wrapper.prop('isBusy')).toEqual(initialState.auth.isBusy);
    })

});
