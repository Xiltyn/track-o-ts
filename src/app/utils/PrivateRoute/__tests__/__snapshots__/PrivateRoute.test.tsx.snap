// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<|:===== <PrivateRoute /> =====:|> renders component 1`] = `
<PrivateRoute
  dispatch={[Function]}
  isAuthenticated={false}
  isBusy={false}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
/>
`;
